test/metabase/upload_test.clj

```diff
@@ -1612,10 +1637,10 @@
 
           (testing "The new row is inserted with the values correctly reordered"
             (is (= {:row-count 1} (append-csv! {:file file, :table-id (:id table)})))
-            (is (= (rows-with-auto-pk
-                    [["Obi-Wan Kenobi" "No one really knows me"]
-                     ["Puke Nightstalker" "Nothing - you can't prove it"]])
-                   (rows-for-table table))))
+            (is (= (set (rows-with-auto-pk
```

Comment:
Why are we need to add this casting to a set, does Clickhouse not provide a stable order?

Not important for this PR, but I think it would be good to test that order is preserved for all the drivers that support that. This probably doesn't warrant a feature or method, but perhaps there's another way to opt out of the stronger test?

Reply:
Without an `order by` clause, Clickhouse queries will return the last rows inserted first, while other drivers return the first rows appended first. I don't see how we can change this without adding an auto-incrementing PK and ordering the results by that column. We don't document any guarantees about the order so I think this is okay for now while we allow drivers to not have auto-incrementing PKs.

Reply:
Ah thanks for clarifying its semantics. Not that it's crucial, if we wanted to make the tests "tighter" in the short term we could only cast to a set if `:upload-with-auto-pk` is false. It would be easy enough to strengthen or relax that internal coupling if things change.